src/openassetio-core-c/private/SimpleMap.cpp

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2013-2022 The Foundry Visionmongers Ltd
+#include <stdexcept>
+
+#include <openassetio/c/SimpleMap.h>
+#include <openassetio/c/StringView.h>
+#include <openassetio/c/errors.h>
+#include <openassetio/SimpleMap.hpp>
+#include <openassetio/typedefs.hpp>
+
+#include "StringView.hpp"
+#include "chandle.hpp"
+#include "errors.hpp"
+
+namespace {
+
+using openassetio::SimpleMap;
+using HandleConverter = openassetio::chandle::Converter<SimpleMap, OPENASSETIO_NS(SimpleMap_h)>;
+
+/**
+ * Get a primitive value from a SimpleMap, converting exceptions to
+ * error codes.
+ *
+ * @tparam Type Type of value to extract from variant.
+ * @param[out] err Storage for error message, if any.
+ * @param[out] out Storage for return value.
+ * @param handle Opaque handle to a SimpleMap.
+ * @param key Key to query within SimpleMap.
+ * @return Error code.
+ */
+template <class Type>
+int get(OPENASSETIO_NS(StringView) * err, Type *out, OPENASSETIO_NS(SimpleMap_h) handle,
+        const OPENASSETIO_NS(ConstStringView) key) {
+  const SimpleMap *simpleMap = HandleConverter::toInstance(handle);
+
+  return openassetio::errors::catchUnknownExceptionAsCode(err, [&] {
+    // TODO(DF): @exception messages.
+    try {
+      *out = std::get<Type>(simpleMap->at({key.data, key.size}));
+    } catch (const std::out_of_range &exc) {
+      // Default exception message:
+      // VS 2019: "invalid unordered_map<K, T> key"
+      // GCC 9: "_Map_base::at"
+      openassetio::assignStringView(err, "Invalid key");
+      return OPENASSETIO_NS(kError_OutOfRange);
+    } catch (const std::bad_variant_access &exc) {
+      // Default exception message:
+      // VS 2019: "bad variant access"
+      // GCC 9: "Unexpected index"
+      openassetio::assignStringView(err, "Invalid value type");
+      return OPENASSETIO_NS(kError_BadVariantAccess);
+    }
+
+    return OPENASSETIO_NS(kOK);
+  });
+}
+}  // namespace
+
+extern "C" {
+
+OPENASSETIO_NS(SimpleMap_s) OPENASSETIO_NS(SimpleMap_suite)() {
```

Comment:
There must be some axiom about things declared "Simple" and computer science oxymorons :) Instead of calling it Simple, why not point directly at its use, which I think might be better served by something like "AttirbuteDictionary" or some such.

Reply:
It's been on our list to rename this and your comment is the straw that broke the camel's back.  So we've decided on `InfoDictionary`.  This sounds rather generic, but the primary purpose for this type is as a return value from an `info` member function. It will also be used as a possible value type received when `resolve`ing an entity reference, but we think the name works for that too, more-or-less.

Reply:
I've gone and rewrote history so it's now (and always was ;) ) called `InfoDictionary`.